### ISSUE TITLE
[discussion] Experiment tracking

### ISSUE THREAD

**Issue by chrisheecho** [kubeflow/kubeflow#264]
Easy way to track your experiments in kubeflow

**Comment by bhack** [comment]
I think this will involve also experiments data managment in  https://github.com/kubeflow/experimental-kvc and model managment in https://github.com/kubeflow/kubeflow/issues/136 /cc @jlewi

**Comment by jlewi** [comment]
Any ideas how we can begin to make progress on this issue?

How should we define an experiment?

Suppose we had an agent in the cluster that would log the spec & status of TFJobs and (possibly other resources) to a DB. So this would give us a record of

- The Docker image used
- The command line arguments used

Would this provide rudimentary experiment tracking?

**Comment by bhack** [comment]
Command line or configmaps? I think we need to add metrics of the experiments and resulting model. Is the datacut defined in configmaps?

**Comment by wbuchwalter** [comment]
There are 3 intesting components in an experiment I think:
1. State of dataset when the experiment was launched
2. Details of the model used
3. Resulting model

`1` is tricky and probably out of scope for now? 
For `2` I think what @jlewi mentionned is enough.
For `3` can we figure out a solution that would work with a wide variety of storage options?

There is also https://github.com/tensorflow/tensorboard/issues/92 but no ETA.

**Comment by bhack** [comment]
I think 1 needs to be partially handled in https://github.com/kubeflow/experimental-kvc

**Comment by jlewi** [comment]
I don't understand the connection between experimental-kvc and 1.  Suppose one is reading/writing data to/from GCS/S3/BigQuery/SQL. Why would experimental-kvc enter the picture?

@wbuchwalter For #3. Isn't the model just a URI?

**Comment by bhack** [comment]
Suppose that you want to reproduce the experiment. You need to find a way to connect experiment with data provisioning. I.e. do you need prepopulation?
Often data provisioning, ETL/Augmentation, Network Trasfer could become a bottleneck for feeding GPU cycles. So I think it is important to track how to provision the data in a modality required to the experiment.

**Comment by bhack** [comment]
Something could be covered by http://www.pachyderm.io/dsbor.html.
So I've added a comment in kvc https://github.com/kubeflow/experimental-kvc/issues/7#issuecomment-375058966.

**Comment by ConnorDoyle** [comment]
We have begun internally testing a prototype for tracking experiments and result metadata as CRDs, together with a lightweight python library to interact with this API from an experiment coordinator (creating k8s objs for points in the parameter space) and the job script itself (emitting results.) So far we have gotten positive feedback from ML practitioners. Happy to share more about the approach, related design decisions, etc.

**Comment by inc0** [comment]
Bumping this issue as it's going to be one of priorities for 1.0 (am I correct @jlewi ?).

So one idea that seems to be most obvious would be model CRD, then Katib/<< our unnamed CI system for models >> could use it to manage model history and update CRD with metrics etc.

**Comment by YujiOshima** [comment]
For Katib, it needs
* Details of Model (parameters, metrics)
* Results (log of metrics)
* Property of Dataset
The log means Stdio of Pods? or saved from TF inside pods?
Does CRD contain the log itself or a path to the log that is stored in PVC?

**Comment by inc0** [comment]
So I suggest Model CRD and operator:

- Model name
- Model path (S3, pvc)
- Storage creds, pvc name, all the details regarding how to access
- Model log path
   - train
   - test
   - inference
- Model metadata
   - hyperparameters etc
   - model metrics (accuracy etc) - filled later with Katib/CI system

Operator could autogenerate paths based on operator specs.
Data scientist would create this custom resource prior to running tfjob, and this would populate env variables of tfjob so data scientist could use it in code.

Thoughts?

**Comment by YujiOshima** [comment]
Cool. All you suggested are needed.
I request two more items.
1. A unique Id for the model.
Users would try training in the same condition. These models should have the same name but another ID like replicas of a pod.

2. A property of input data.

* Dataset metadata
   * Dataset name (cifar10, ImageNet etc)
   * Dataset Path

**Comment by bhack** [comment]
Also a short fulltext description of the exp

**Comment by inc0** [comment]
For dataset I was thinking of integration with pachyderm and let it manage data versioning etc, I'll reach out to them and see how can we merge these.

**Comment by dwhitena** [comment]
We can definitely version data and connect that with processing stages in Pachyderm. We also automatically track the Docker images, commands, stdin, env vars, etc. for each processing stage along with it's input/output data to get full "data provenance" of experiments.

Out of the things mentioned by @wbuchwalter:

```
1. State of dataset when the experiment was launched
2. Details of the model used
3. Resulting model
```

Pachyderm solves (1) and (3) completely as in. (2) is partially solved in that you can version parameters, configuration, etc. as long as those are represented in data or command line arguments. The part that Pachyderm doesn't solve in a transparent way is the versioning of model internals/architecture, although that is versioned in model binaries and parameter files (at least partially).

The only downside of this is that the pipeline "specifications" are in a Pachyderm specific format: http://pachyderm.readthedocs.io/en/latest/reference/pipeline_spec.html. If you wanted a different format, maybe there could be some automated transformation or export.

**Comment by dwhitena** [comment]
To follow up on the above, @inc0 mentioned the following in Slack:

> well for us we'll probably need more metadata - path to actual model being saved somewhere (S3), hyperparams used, path to logs etc

In this regard, it really depends on what format you need. All of those things are available in pachyderm, but I think you could create a layer on top that would bundle them in a nice way for consumption (which might also involve translating pachyderm pipeline specs, exposing S3 paths, etc.).

**Comment by nqn** [comment]
Wanted to add that we have a few users on this project: https://github.com/IntelAI/experiments

**Comment by inc0** [comment]
so, I think we all agree that this piece is needed so, if nobody objects, I'll go ahead and request repo for it.

**Comment by YujiOshima** [comment]
@inc0 Of course I agree! We can discuss at the new repo. I will join it!

**Comment by jlewi** [comment]
Can we have an issue or proposal providing some insight into what this repo will cover? **Experiment tracking** is rather vague.

Perhaps we can start by defining experiment?

The repository linked to above by @nqn defines experiment as 

> Experiment Describes a hyperparameter space and how to launch a job for a sample in that space. Has a unique name.

To me this sounds exactly like what Katib is doing so if that's what we mean by experiment its not clear to me why we would need a new repository as opposed to using Katib.

**Comment by bhack** [comment]
What I think is that an hyperparamter could be consider part of the config(map)/args changes but often you change the sourcecode and so the image for an experiment. So for me an experiment track also source/image changes cause it is up to the user to choice for an experiment what is a part of the config and what is part of the source code cause the first don't require an new image.

**Comment by bhack** [comment]
I generally think that an experiment can change across three axes: 
- source code changes: it needs git hash/image tracking
-  config(maps) changes: it needs config tracking
- data input changes: need input tracking/versioning

**Comment by YujiOshima** [comment]
@inc0 
As @jlewi said, Katib store almost all items you suggest in DB now.
You can save and get them with Katib API.
Please make the scope of new repo clearly.
Define CRD items? or make new operator?
If only define CRD items, you can discuss at Katib repo.
If you will make new operator, what the difference to existing operator?

**Comment by jlewi** [comment]
> Now this CRD + Operator (for doing things like "retrain" or "spawn serving cluster") needs to live somewhere. We can either create new repo, which I think makes sense since this will be useful for all other pieces of Kubeflow really, or put it into Katib and therefore extend it's scope quite a bit.

Why do we need a new CRD for serving? SeldonIO already provides a CRD for managing model deployments.

Since Katib provides a DB that tracks models and their performance, I assume if one wanted to deploy a model you would look in Katib to pick model and then create a Seldon Deployment or K8s deployment for TFServing and then just deploy that.

**Comment by inc0** [comment]
So I'm ok with storing this code in Katib, it was my understanding before that Katib itself didn't want this scope.
I wouldn't want to use Katib db, I'd prefere CRD because it allows us to make operator work, reuse this resource in different places (serving being my main focus atm) etc. That would also expand scope of Katib quite a bit, since it's more than just hyperparameter tuning.

I see 2 solutions:
1. We create new repo, CRD, logic etc and Katib uses it rather than internal db
OR
2. We make Katib source of truth and expand its scope to model tracking and infrastructure.

My main goal in this work is to provide easy way to manage existing models, upgrading them, retraining them, serving them etc. That requires keeping track of them, which part of Kubeflow does that, I'm not sure. I'm feel quite strongly that we should make this CRD tho...

**Comment by inc0** [comment]
@jlewi Seldon might, tf-serving doesn't. And what about CI system for upgrades etc? If we say that we only support Seldon for things like upgrading model, load balancing model etc etc, that's quite a decision.

**Comment by jlewi** [comment]
Its not clear to me what problem you are trying to solve. Can you provide a concrete example (perhaps you could use our GitHub issue summarization example as the scenario), and explain what you want to do, how you would do it today (if possible) and how you want to improve the experience.

**Comment by inc0** [comment]
Ok, let's use github issue summarization. Workflow for user would look like this:

* create model CRD, with initial info like path in S3 and so on
* spawn TFjob with these parameters 
* tfjob uploads actual model file to proper S3 space
* Run CI scripts to populate metrics
* Spawn serving cluster based on these information

... After new data is available

* Copy model resource and retrain model with new data, possibly using old one as initialization
* Possibly modify hyperparameters and retrain
* Populate new models CI metrics (like new accuracy)
* if it's better than old one, perform upgrade on prod

All of these ^ will need some record of model. Only one step is related to Katib and it's solved by internal db rather than resource that can be reused.

**Comment by jlewi** [comment]
So what's the pain point in the above workflow you are trying to address?

@YujiOshima Can't we just store the metrics in Katib? I assume we would just add a row in the DB with the accuracy metrics and a link to the path of the model.

> Copy model resource and retrain model with new data, possibly using old one as initialization

What is model resource? The model is trained with a TFJob. So to retrain we just launch another TFJob with appropriate arguments.

**Comment by YujiOshima** [comment]
We can store metrices in Katib. The format of model infomation in Katib is

```
ModelInfo {
     string study_name (experiment name)
     string worker_id ( id of model )
     list Parameters
	-
	string parameter_name
	string value
	enum parameter_type //int, float, categorical

     list metrics
	-
	string metrics_name
	float32 value
	
     string model_path
 }
```

You can update the metrics values any time.

**Comment by inc0** [comment]
Point I'm making is if it's internal to Katib, then we'll need to provide way for other components to access it, but also authorization to db, replication for HA, API etc etc.
If we use CRD it all will be handled by kubernetes.
At some point we'll need to handle multitenancy and access control and more we leverage k8s for that the better.
Again I'm fine with CRD living in Katib, if we extend scope of Katib for this, but I think we need CRD.

@jlewi all of them since all of them need information about model. This is backend thing.

**Comment by cliveseldon** [comment]
My initial thoughts are its not clear why existing components don't already solve some of these proposals. For example:
 * Pachyderm for model data lineage
 * Seldon's CRD or TFServing for deploying/rolling updates to serving models
 * Argo and or Pachyderm for tieing workflow steps together

Some of the other discussion then seems more about how to combine these into a CI/CD pipelines but in a way to allow reproducibility. I think one idea might be to follow "gitops" ideas to allow one to go to a particular commit and rerun everything. 

I agree metrics/logs need to go somewhere to be collected related to one model/experiment but this should generally be a pluggable component where people should have some choice. Forgive me if I have missed the latest kubeflow proposal on that or is this the core concern here?

**Comment by inc0** [comment]
There need to be track of artifacts coming from training tho - for CI, relearn with same parameters but new data, learn from existing checkpoint when new data is available etc etc.
Just pushing code is not enough to trigger deployment of new model (we need to fully train it before we get proper metrics regarding it's performance) and there are more triggers to deploy new model (new data available->retrain), so there needs to be intermediate step in between.
For all that we need metadata of model, like it's location. Proposal is to add CRD to hold this information.

**Comment by bhack** [comment]
I.e. Just to mention the case of TF operator and TF high level API there is a specific arg/configmap to be handled related to [the warm start](https://www.tensorflow.org/versions/master/api_docs/python/tf/estimator/WarmStartSettings)

**Comment by chrisheecho** [comment]
Not today's but next Community call, I would like to propose couple problems that we can look at as a community, decide on one or two problems to try -> then agree on which architecture choices we want to make if possible. does anyone have any objections?

**Comment by jlewi** [comment]
Is there broad agreement about model metadata? e.g. what the API/ Spec should like? Lacking agreement on a common format for model metadata its not clear how different components could begin to work together without requiring the user to write specific glue code.

Also I don't understand why the operational complexity of an SQL database is a factor here. Most clouds have managed offerings and I suspect there are reasonable on prem offerings that most customers are already running.

My expectation is that open service broker provides an abstraction layer which will allow Kubeflow to request a DB without being concerned about the underlying implementation.

**Comment by inc0** [comment]
I think schema is not (and shouldn't be for some time, at least until 1.0) clad in stone. I think this should be extensible anyway.

**Comment by JoeyZwicker** [comment]
@inc0 Sorry if this isn't relevant as I'm coming in a bit late to the conversation, but isn't
> There need to be track of artifacts coming from training tho - for CI, relearn with same parameters but new data, learn from existing checkpoint when new data is available etc etc.

solvable, as @cliveseldon mentioned, by Pachyderm's data version and lineage tracking features? We can retrain with same parameters and new data, same data and new parameters, as well as new code in any of these situations. Pachyderm's incremental features make it really easy to pick up from a checkpoint and continue training your model incrementally as new data is added to the training set.

**Comment by inc0** [comment]
@JoeyZwicker I'm not so much concerned about tracking of data, as it's solvable with pachyderm, but more about tracking of models. Models will be produced from versioned data (link to pachyderm), versioned code (link to tfjob image) and versioned hyperparameters (link katib or metadata in model CRD). On top of that you need metadata like path to actual model file, results of CI runs against this model etc. I don't think all of this is within Pachyderm scope.

**Comment by JoeyZwicker** [comment]
Pachyderm does actually version the output model binaries and parameter files as well and with our provenance feature, you can link parameters + training data + code (via tfj image) to output models and any metadata outputs (labels, annotations, etc). 

For our ML use cases, we generally have users set parameters as data like in our [hyperparameter example here](https://github.com/pachyderm/pachyderm/tree/master/doc/examples/ml/hyperparameter). This is of course a simple demo example, but it's very powerful because it allows you to make parameter changes and version them just like data changes and have those adjustments propagate new jobs downstream with full provenance tracking.

**Comment by YujiOshima** [comment]
Pachyderm looks so useful.
Can Pachyderm manage other data schema than that presented in this example?
Katib needs to store many kinds of data that are specialized for Katib e.g. Study, Trial.
If it can, it will be able to be used in Katib as a backend DB.
But I think Katib doesn't need to depend on it deeply.
We can prepare plugins for MySQL and Pachyderm and users can switch them.

**Comment by JoeyZwicker** [comment]
Pachyderm is just a file system interface so it works with any data type/schema -- text, images, binary files, csv, etc. Keep in mind that _is_ a file-based system and our versioning is at the per-file level. So if your data is a giant table or DB dump, you'll want to break that up into smaller files (we have some sugar cmds to help with this) so you get more efficient diffing and incremental processing. 

@YujiOshima Does that answer your question? I'm only familiar with Katib at a high level so what does a study or trial "kind of data" mean in Katib?

**Comment by YujiOshima** [comment]
@JoeyZwicker I'm sorry for late reply. Great, it is not so big DB table.
They are defined [here](https://github.com/kubeflow/katib/blob/master/pkg/db/db_init.go).
I believe they can be stored in Pachyderm.

@inc0 In you mean the data of experiments and models will be shared by Pachyderm among any components?
Then we can discuss data schema based on Katib's.
And I think Katib can use Pachyderm as the backend DB.

**Comment by inc0** [comment]
@YujiOshima also sorry for late reply:) (I'm super busy with infrastructure stuff now, but I'll get back to Katib as soon as I get some breathing room).

So I see 2 potential avenues to bind Katib and Pachyderm (using CRD):
1. We use Pachyderm's native CRDs, Model then become just another piece of "data" that would fit into pachyderm pipeline. I'm not sure how far can we stretch it, I'll play with it
2. Create our own Model CRD that will also bind with Pachyderms and use same S3 as Pachyderm underneath to hold model BLOB. That would allow easy translation of what today is database in Katib to use "k8s" backend. I'll try to PoC it as well. I'm personally in favor of this option as it feels more "kubernetesy" to me and allows further extensions easily.

**Comment by YujiOshima** [comment]
@inc0 Thank you! I also prefer 2nd option.
I'm looking forward to your PoC!

**Comment by bhack** [comment]
Has anybody tought about experiments tracking and gitops?

**Comment by inc0** [comment]
@bhack It's going to be quite easy with CRDs, operators are well fit for this flow. I'm not sure if we'd start with this but it'd make cool demo for sure:)

**Comment by bhack** [comment]
@inc0 Good! I was looking a little bit at the [kustomize idea of workflow](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/workflows.md)

**Comment by bhack** [comment]
@inc0 there is also another thread for gitops in kubeflow at https://github.com/kubeflow/kubeflow/issues/971 if you are interested.

**Comment by jlewi** [comment]
kubeflow/examples#134 has some early experiments trying to load model data into modeldb for experiment tracking.

**Comment by bhack** [comment]
@jlewi Is there [something moving](https://github.com/tensorflow/tensorboard/pull/1281) on TB side?

**Comment by jlewi** [comment]
If there's I haven't heard about it.
